### PR TITLE
einsum

### PR DESCRIPTION
**Pull request by fukatani**
I tried to port `einsum` from tensorflow.
It seems work well.
If porting from tensorflow is acceptable, I will refactor for cupy and add test.

Part of `numpy.einsum` function which is not supported by tensorflow is not scope of this PR.

#19

### CODE REVIEW

**Comment by unnonouno** [comment]
It sounds great. We need to consider its license.

**Comment by fukatani** [comment]
Another option is constructing einsum from full scratch based on numpy implementation.
It may be better for us to support same function as numpy.
So I'm inspecting numpy implementation.

**Comment by jekbradbury** [comment]
PyTorch is also writing einsum right now https://github.com/pytorch/pytorch/pull/2008/files, and they were also unsure about porting TensorFlow code and decided not to; the PyTorch license should not cause any problems for you though.

**Comment by fukatani** [comment]
I'm writing `einsum` from scratch.
This is because I thought that writing with full scratch would be better, not licence problem.

I mean einsum can be more concise and powerful than tensorflow's.
ex. support `diagonal` functionality and case three or more operands have same label.

I'll start to write doc and process exception.

File: cupy/linalg/einsum.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+    arrow_pos = subscripts.find('->')
+    if arrow_pos == -1:
+        input_subscripts = subscripts
+        label_list = list(input_subscripts.replace(',', ''))
+        out_label_set = set(label_list) - get_dummy_labels(label_list)
+        output_subscript = ''.join(sorted(list(out_label_set)))
+    else:
+        input_subscripts = subscripts[:arrow_pos]
+        output_subscript = subscripts[arrow_pos + 2:]
```
Comment (unnonouno) [inline cupy/linalg/einsum.py]: Does it correctly raise an error when ouput_subscript includes "," such as "ij->i,j"?

Comment (unnonouno) [inline cupy/linalg/einsum.py]: Why don't you use regexp to parse subscripts?

File: cupy/linalg/einsum.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        cupy.ndarray:
+            The calculation based on the Einstein summation convention.
+
+    .. seealso:: :func:`numpy.einsum`
+    """
+
+    # TODO(fukatani): Support 'out', 'order', 'dtype', 'casting', 'optimize'
+
+    if not operands:
+        raise ValueError("must specify the einstein sum subscripts string and "
```
Comment (fukatani) [inline cupy/linalg/einsum.py]: If we use `def einsum(subscripts, *operands):`, `cupy.einsum()` raises `TypeError` instead of `ValueError`.

`numpy.einsum` also uses `def einsum(*operands):`.
See 
https://github.com/numpy/numpy/blob/e3b5e88ae6c0b546dd63b869f32427553ad36e57/numpy/core/einsumfunc.py#L813

**Comment by fukatani** [comment]
Thanks! I addressed review comments excepts `def einsum(subscripts, *operands):`.

File: cupy/linalg/einsum.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+def calc_single_view(ioperand, subscript):
+    """Calculate 'ii->i' by cupy.diagonal if needed.
+
+    Args:
+        ioperand (cupy.ndarray): Array to be calculated diagonal.
+        subscript (str):
+            Specifies the subscripts. If the same label appears
+            more than once, calculate diagonal for those axes.
+    """
+
```
Comment (unnonouno) [inline cupy/linalg/einsum.py]: Can you add more assertion to understand its spec?
```
assert ioperand.ndim == len(subscript)
```
?

File: cupy/linalg/einsum.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    Args:
+        ioperand (cupy.ndarray): Array to be summed.
+        input_subscript (str): Specifies the subscripts for input array.
+        output_subscript (str):
+            Specifies the subscripts for output array. If one label exists in
+            input_subscript but not in output_subscript, this label will be
+            summed.
+    """
+
```
Comment (unnonouno) [inline cupy/linalg/einsum.py]: ditto
```
assert len(set(input_subscript)) == len(input_subscript)
assert len(set(output_subscript)) == len(output_subscript)
assert set(output_subscript).issubset(set(input_subscript))
```
?

File: cupy/linalg/einsum.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    Args:
+        ioperand (cupy.ndarray): Array to be transpose.
+        input_subscript (str): Specifies the subscripts for input arrays.
+        output_subscript (str):
+            Specifies the subscripts for output arrays. If input does not
+            match output, ``operand`` is transposed so that it matches.
+    """
+
+    assert len(input_subscript) == len(output_subscript)
```
Comment (unnonouno) [inline cupy/linalg/einsum.py]: Check uniqueness instead.
```
assert len(set(output_subscript)) == len(output_subscript)
```

**Comment by unnonouno** [comment]
jenkins test this please

File: cupy/linalg/einsum.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+    converted_inputs = []
+    dtype = numpy.result_type(*ioperands)
+    for a in ioperands:
+        if isinstance(a, cupy.ndarray):
+            converted_inputs.append(a.astype(dtype))
+        else:
+            converted_inputs.append(cupy.asarray(a, dtype=dtype))
+
+    match = re.fullmatch('([a-zA-Z,]+)(->[a-zA-Z]*)?', subscripts)
```
Comment (unnonouno) [inline cupy/linalg/einsum.py]: `fullmatch` is only supported in py3. Please use `^` and `$` for py2.

**Comment by fukatani** [comment]
Thank you for your attentive review!
I addressed your comments.

By the way, I copied doc from numpy. Should I added any comments?

**Comment by unnonouno** [comment]
It is OK. The license contains numpy's one. https://github.com/cupy/cupy/blob/master/LICENSE#L25

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
LGTM! It is great work!!
